wurst/file/FileIO.wurst

```diff
@@ -0,0 +1,105 @@
+package FileIO
+import public ChunkedString
+import ErrorHandling
+import ObjectIdGenerator
+import AbilityObjEditing
+import ObjectIds
+
+/* 	FileIO for Patches 1.31 and up with chunked string support for big data.
+	Reads and writes using preload files by manipulating default ability tooltips.
+
+	Read a file:
+	>	let fileContent = new File("MyFileName.pld").readAsString()
+
+	Write a file:
+	>	new File("MyFileName.pld")..write("MyContent")..close()
+*/
+
+@configurable public constant CHUNKS_PER_FILE = 64
+
+public var CAN_READ_FILES = true
+
+constant DUMMY_ABIL_ID = compiletime(ABIL_ID_GEN.next())
+
+public class File
+	var filename = ""
+
+	construct(string filename)
+		this.filename = filename
+
+	function write(player p, string content)
+		let buffer = new ChunkedString()
+		buffer.append(content)
+		write(p, buffer)
+		destroy buffer
+
+	function write(player p, ChunkedString buffer)
+		for i = 0 to buffer.chunkCount
+			let str = buffer.getChunk(i)
+			if validateInput(str) != null
+				error("FileIO(" + filename + ") ERROR: Invalid character |cffffcc00" + validateInput(str) + "|r")
+
+		writePreload(p, buffer)
+
+	function close()
+		destroy this
+
+	function read(player p) returns ChunkedString
+		let out = new ChunkedString(DEFAULT_CHUNK_SIZE)
+		if localPlayer == p
+			readPreload(out)
+		return out
+
+	function readAsString(player p) returns string
+		let out = new ChunkedString(DEFAULT_CHUNK_SIZE)
+		if localPlayer == p
+			readPreload(out)
+		return out.readChunk()
+
+	private function writePreload(player p, ChunkedString buffer)
+		if localPlayer == p
+			PreloadGenClear()
+			PreloadGenStart()
+
+			for i = 0 to buffer.chunkCount
+				if (i >= CHUNKS_PER_FILE)
+					error("FileIO(" + filename + ") ERROR: String exceeds max length (" + (CHUNKS_PER_FILE * buffer.chunkSize).toString() + ").|r")
+
+				let chunk = buffer.getChunk(i)
+				Preload("\" )\ncall BlzSetAbilityTooltip('" + DUMMY_ABIL_ID.toRawCode() + "', \"" + chunk + "\", " + i.toString() + ")\n//")
```

Comment:
You may want to precompute `toRawCode()`, since writing files (especially big) may be performance sensitive, and `toRawCode` does a loop with divisions inside, and here it's used in a loop, or is the optimizer able to detect it and do it itself, since the `DUMMY_ABIL_ID` is a compiletime value?

Reply:
the optimizer happens after compiletime stuff is over. I don't think toRawCode would bottleneck here, but I added the global.